src/imitation/density_baselines.py

```diff
@@ -0,0 +1,267 @@
+          # hierarchical model in which we first check whether state is
+          # absorbing, then assign either constant score or a score based on
+          # density.
+          raise Exception(
+              f"Time {time} out of range (0, {len(self._density_models)}], and "
+              f"I haven't implemented absorbing states etc. yet")
+        else:
+          time_model = self._density_models[time]
+          rew = time_model.score(scaled_padded_trans)
+      rew_list.append(rew)
+    rew_array = np.asarray(rew_list, dtype='float32')
+    return rew_array
+
+
+class DensityTrainer:
+  def __init__(self,
+               env,
+               rollouts,
+               imitation_trainer,
+               *,
+               standardise_inputs=True,
+               kernel='gaussian',
+               kernel_bandwidth=0.5,
+               density_type=STATE_ACTION_DENSITY,
+               is_stationary=False):
+    r"""Family of simple imitation learning baseline algorithms that apply RL to
+    maximise a rough density estimate of the demonstration trajectories.
+    Specifically, it constructs a non-parametric estimate of `p(s)`, `p(s,s')`,
+    `p_t(s,a)`, etc. (depending on options), then rewards the imitation learner
+    with `r_t(s,a,s')=\log p_t(s,a,s')` (or `\log p(s,s')`, or whatever the
+    user wants the model to condition on).
+
+    Args:
+      env (gym.Env or str): environment to train on.
+      imitation_trainer (BaseRLModel): RL algorithm & initial policy that will
+        be used to train the imitation learner.
+      expert_trainer (BaseRLModel): snapshot of RL algorithm that has been
+        trained to solve `env`.
+      n_expert_trajectories (int): number of expert trajectories to sample and
+        use as demonstrations.
+      kernel, kernel_bandwidth, density_type, is_stationary,
+        n_expert_trajectories: these are passed directly to `DensityReward`;
+        refer to documentation for that class."""
+    self.env = util.maybe_load_env(env, vectorize=True)
+    self.imitation_trainer = imitation_trainer
+    self.reward_fn = DensityReward(trajectories=rollouts,
+                                   density_type=density_type,
+                                   obs_space=self.env.observation_space,
+                                   act_space=self.env.action_space,
+                                   is_stationary=is_stationary,
+                                   kernel=kernel,
+                                   kernel_bandwidth=kernel_bandwidth,
+                                   standardise_inputs=standardise_inputs)
+    self.wrapped_env = reward_wrapper.RewardVecEnvWrapper(self.env,
+                                                          self.reward_fn,
+                                                          include_steps=True)
+    self.graph = tf.Graph()
+    self.sess = tf.Session(graph=self.graph)
+    with self.graph.as_default():
+      self.sess.run(tf.global_variables_initializer())
```

Comment:
[Else](https://github.com/HumanCompatibleAI/imitation/blob/a0f2f7b88212851058ba6224b60750fae01d8c4d/src/imitation/trainer.py#L77)[where](https://github.com/HumanCompatibleAI/imitation/blob/5d775e7bac886a067bc4c08faa9b4701a7a26164/src/imitation/discrim_net.py#L18) in the repo we expect a default session to exist.

We haven't decided which of the two tf.Session strategies that @AdamGleave described #31 we are going for, so it's okay to leave this for now (fixing it later should be easy).